### PR TITLE
gRPC proxying of image registry requests

### PR DESCRIPTION
**Pull request by mikejholly**
This PR adds support for proxying image registry requests to the BK-embedded registry using a gRPC stream. A local proxy is used by the container frontend to "pull" an image from the local or remote BK via the gRPC proxy.

image

Associated with: https://github.com/earthly/buildkit/pull/25

### CODE REVIEW

File: cmd/earthly/subcmd/build_cmd.go [hunk 1]
```diff
@@ -579,14 +583,48 @@ func (a *Build) ActionBuildImp(cliCtx *cli.Context, flagArgs, nonFlagArgs []stri
...
+			err := regProxy.Serve(ln)
+			if err != nil && !errors.Is(err, http.ErrServerClosed) {
+				a.cli.Console().Warnf("Failed to serve registry proxy: %v", err)
+			}
+		}()
+
+		defer func() {
+			ctx := context.Background()
+			err := regProxy.Shutdown(ctx)
+			if err != nil {
```
Comment (vladaionescu) [inline cmd/earthly/subcmd/build_cmd.go:622]: Maybe don't warn if the err is a cancellation.

**Review by vladaionescu — COMMENTED** [review]
Looks good overall! Left some small comments below...

It would be good if we can add a test for this. We should be able to reuse [this script](https://github.com/earthly/earthly/blob/main/.github/workflows/reusable-export-test.yml) entirely. This is called from this [CI yaml](https://github.com/earthly/earthly/blob/main/.github/workflows/reusable-export-test.yml). I guess all we have to do is create a new CI job that runs this existing test with the right flag turned on.

File: cmd/earthly/subcmd/build_cmd.go [hunk 2]
```diff
@@ -579,14 +583,48 @@ func (a *Build) ActionBuildImp(cliCtx *cli.Context, flagArgs, nonFlagArgs []stri
...
+
+		go func() {
+			err := regProxy.Serve(ln)
+			if err != nil && !errors.Is(err, http.ErrServerClosed) {
+				a.cli.Console().Warnf("Failed to serve registry proxy: %v", err)
+			}
+		}()
+
+		defer func() {
+			ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (mikejholly) [reply cmd/earthly/subcmd/build_cmd.go:620]: Thanks, I didn't notice that var.

File: features/features.go [hunk 3]
```diff
@@ -60,6 +60,7 @@ type Features struct {
 	EarthlyCIRunnerArg         bool `long:"earthly-ci-runner-arg" description:"includes EARTHLY_CI_RUNNER ARG"`
 	UseDockerIgnore            bool `long:"use-docker-ignore" description:"fallback to .dockerignore incase .earthlyignore or .earthignore do not exist in a local \"FROM DOCKERFILE\" target"`
 	PassArgs                   bool `long:"pass-args" description:"Allow the use of the --pass-arg flag in FROM, BUILD, COPY, WITH DOCKER, and DO commands"`
+	UseRemoteRegistry          bool `long:"use-remote-registry" description:"Make use of the embedded BuildKit Docker registry for image pulls (rather than Tar transfer)"`
```
[the remark this answers is not in the document]
Comment (mikejholly) [reply features/features.go:63]: No, I added it initially, but it seems like a flag may be a better fit as the features aren't really available until we parse an Earthfile. Was looking for some feedback on the most appropriate strategy here.

File: cmd/earthly/subcmd/build_cmd.go (hunk 1 above)
Comment (mikejholly) [reply cmd/earthly/subcmd/build_cmd.go:622]: Good call. Thanks.

File: cmd/earthly/flag/global.go [hunk 4]
```diff
@@ -510,5 +509,12 @@ func (global *Global) RootFlags(installName string, bkImage string) []cli.Flag {
 			Usage:       "A remote docker image tag use as explicit cache",
 			Destination: &global.RemoteCache,
 		},
+		&cli.BoolFlag{
+			Name:        "use-remote-registry",
+			EnvVars:     []string{"EARTHLY_USE_REMOTE_REGISTRY"},
+			Usage:       "Use remote registry with Docker pull",
+			Destination: &global.UseRemoteRegistry,
+			Hidden:      true, // Internal.
+		},
```
[the remark this answers is not in the document]
Comment (mikejholly) [reply cmd/earthly/flag/global.go:518]: Good idea :+1:

File: buildkitd/Earthfile [hunk 5]
```diff
@@ -12,7 +12,7 @@ buildkitd:
             ARG BUILDKIT_BASE_IMAGE=$BUILDKIT_PROJECT+build
         END
     ELSE
-        ARG BUILDKIT_BASE_IMAGE=github.com/earthly/buildkit:8df7c52e603ee6db6a471fcb7f6a85a407f611f4+build
+        ARG BUILDKIT_BASE_IMAGE=github.com/earthly/buildkit:40cd44f81962f99200f10df3a79d044f9df0aefe+build
```
Comment (alexcb) [inline buildkitd/Earthfile:15]: don't forget to change this to the rebased commit :)

File: features/features.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vladaionescu) [reply features/features.go]: I think that the CLI flag makes sense for this. Usually, we use the features.go flags for syntax changes.

**Review by vladaionescu — APPROVED** [review]
👏 👏

File: buildkitd/Earthfile (hunk 5 above)
Comment (alexcb) [reply buildkitd/Earthfile]: this still needs updating, earthly-main under buildkit is now at `af796f0cb8ff6cf53fdebd543ff1e7cd173c00a1`

**Review by alexcb — APPROVED** [review]
looks good, only nit is related to pining to the pre-merged commit of buildkit.

File: buildkitd/Earthfile (hunk 5 above)
Comment (mikejholly) [reply buildkitd/Earthfile]: Thanks. Updated to `af796f0cb8ff6cf53fdebd543ff1e7cd173c00a1`

**Review by alexcb — APPROVED** [review]
thanks!

**Pull request merged**